pkg/pingpong/pingpong.go

```diff
@@ -95,9 +99,15 @@ func (s *Service) Ping(ctx context.Context, address swarm.Address, msgs ...strin
 	return time.Since(start), nil
 }
 
-func (s *Service) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
+func (s *Service) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) (err error) {
 	w, r := protobuf.NewWriterAndReader(stream)
-	defer stream.FullClose()
+	defer func() {
+		if err != nil {
+			_ = stream.Reset()
```

Comment:
same here, with an additional missed edge case that the error may be EOF, in which case the loop would exit and the stream would erroneously be reset (it shouldn't).

Reply:
on EOF we break and return nil, so no reset happens.

isn't the reset effectual in a sense that it avoids a dangling-until-timeout read on a stream while the other side is not writing because it encountered some error?

Reply:
Reset is used to indicate to the other peer that there was an error. actually at a second look, the EOF won't be considered as an error since an explicit nil is returned. In any case, this reset is ineffectual since there is no point of resetting an errored stream. I don't know how libp2p propagates errored streams across nodes but I won't jump into any conclusions without a careful inspection of their source code.